src/Compiler/Service/TransparentCompiler.fs

```diff
@@ -1443,115 +1443,141 @@ type internal TransparentCompiler
         FSharpParseFileResults(diagnostics, parseTree, true, [||])
 
     let ComputeParseAndCheckFileInProject (fileName: string) (projectSnapshot: ProjectSnapshot) =
-        caches.ParseAndCheckFileInProject.Get(
-            projectSnapshot.FileKey fileName,
-            node {
-
-                use _ =
-                    Activity.start "ComputeParseAndCheckFileInProject" [| Activity.Tags.fileName, fileName |> Path.GetFileName |]
-
-                match! ComputeBootstrapInfo projectSnapshot with
-                | None, creationDiags -> return emptyParseResult fileName creationDiags, FSharpCheckFileAnswer.Aborted
-
-                | Some bootstrapInfo, creationDiags ->
-
-                    let priorSnapshot = projectSnapshot.UpTo fileName
-                    let! snapshotWithSources = LoadSources bootstrapInfo priorSnapshot
-                    let file = snapshotWithSources.SourceFiles |> List.last
-
-                    let! parseResults = getParseResult projectSnapshot Seq.empty file bootstrapInfo.TcConfig
-
-                    let! result, tcInfo = ComputeTcLastFile bootstrapInfo snapshotWithSources
-
-                    let (tcEnv, _topAttribs, checkedImplFileOpt, ccuSigForFile) = result
-
-                    let tcState = tcInfo.tcState
-
-                    let sink = tcInfo.sink.Head // TODO: don't use head
-
-                    let tcResolutions = sink.GetResolutions()
-                    let tcSymbolUses = sink.GetSymbolUses()
-                    let tcOpenDeclarations = sink.GetOpenDeclarations()
-
-                    // TODO: Apparently creating diagnostics can produce further diagnostics. So let's capture those too. Hopefully there is a more elegant solution...
-                    // Probably diagnostics need to be evaluated during typecheck anyway for proper formatting, which might take care of this too.
-                    let extraLogger = CapturingDiagnosticsLogger("DiagnosticsWhileCreatingDiagnostics")
-                    use _ = new CompilationGlobalsScope(extraLogger, BuildPhase.TypeCheck)
+        node {
+            let fileKey = projectSnapshot.FileKey fileName
 
-                    // Apply nowarns to tcConfig (may generate errors, so ensure diagnosticsLogger is installed)
-                    let tcConfig =
-                        ApplyNoWarnsToTcConfig(bootstrapInfo.TcConfig, parseResults.ParseTree, Path.GetDirectoryName fileName)
+            let file =
+                projectSnapshot.SourceFiles |> List.tryFind (fun f -> f.FileName = fileName)
 
-                    let diagnosticsOptions = tcConfig.diagnosticsOptions
+            match file with
+            | None -> return emptyParseResult fileName [||], FSharpCheckFileAnswer.Aborted
+            | Some f ->
+                let! source = f.GetSource() |> NodeCode.AwaitTask
```

Comment:
Hmm, now I see having to load the file's source makes this a bit complicated. Could we instead rely on the `Version` that is given in the snapshot?

Reply:
Mmh, you mean the FSharpFileSnapshot.Version?
I'm not sure this changes when just the source code of the file changes.

Reply:
Yeah. It definitely should change, otherwise we don't even try to look at the source and assume it's the same as before. 

(Only in VS the version also changes when the file is just open, even when source didn't change.)

Reply:
Okay, seems like we can make this work just with the ProjectSnapshot.FileKey, as long as that changes between file edits.

Reply:
That will change for sure. But it will also change when any of the previous files in the project or any dependencies change. 

That's why I thought we can add file `Version` here:
https://github.com/dotnet/fsharp/blob/9d36d648f85fa5acc6951b0504d538f24f48345e/src/Compiler/Service/FSharpProjectSnapshot.fs#L238

And then look it up the same way we did with the hash - take any result as long the file version is the same.  

So when we have previous results which match the file contents we get them, even if some prerequisites have changed. Which I thought is what this should do. 

If we want it to match only if the result is for sure valid including all dependencies than it could be simplified even further, but I don't think that's the goal of this.

Reply:
So you want to have something like
`member _.GetVersion() = hash |> Md5Hasher.addBytes f.Version |> Md5Hasher.toString` ?

I thought about that but the if condition in `lastFileHash` above made me think that's not the right thing to do:
https://github.com/dotnet/fsharp/blob/9d36d648f85fa5acc6951b0504d538f24f48345e/src/Compiler/Service/FSharpProjectSnapshot.fs#L225-L228

Reply:
No, I meant that it would be a tuple of `(whateverIsThereNow, FileVersion)` so that you can filter it out using the predicate function based just on the file version.

The file version is already in the hash, you just can't match on it alone. 

Sorry this is confusing, we probably need some better vocabulary for the various _key_ and _version_ usages 🙄

Reply:
Sorry, yeah, I'm really confused by now.

I tried `member _.GetVersion() = hash |> Md5Hasher.toString, f.Version` but that changes all the caches that currently use  `ProjectSnapshot.FileKey` as their key. Their tests start failing because caching behaviour changes.
So maybe it's a bit too much to force this into the Version of `FileKey`..

Reply:
Huh, I thought it wouldn't change anything because the version should already be in the hash 🤔. I'll try to have a look why that is.

But in that case we could just make a custom ICacheKey for this one cache. It would be like you had with the hash, but it won't have to be async and can be a method/property on the snapshot maybe.

Reply:
Just do double check - you'd need to do `f.Version |> Md5Hasher.toString` because here the version is `byte array` which won't match anything but the same instance, so can't directly be used.

Reply:
I tried adding it locally and tests seem to be passing. Still not really sure whether it's a good idea to add it to all the other caches where we're not actually going to use it. In theory it could have some minuscule performance impact. So maybe a separate key is better anyway.

Reply:
No worries, pretty sure I fell into the same trap before 😄. The byte arrays are kinda ugly but at least it's reasonably fast that way.